File: mobile/ios/Air/SubModules/WalletContext/Helpers/TmailHelpers.swift
//
//  TmailHelpers.swift
//  MyTonWalletAir
//
//  Tmail alias detection helpers. Kept separate from DNSHelpers so the two detectors don't mix.
//

import Foundation

private let TMAIL_DOMAIN_SUFFIX = "@tmail.ton"
private let TMAIL_SHARE_PATH = "/share/"
private let TMAIL_ALIAS_REGEX = try! NSRegularExpression(
    pattern: "^[a-z0-9]([-_+a-z0-9]{0,62}[a-z0-9])?$",
    options: .caseInsensitive
)

public class TmailHelpers {
    private init() {}

    public static func isTmailAlias(_ value: String) -> Bool {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard trimmed.hasSuffix(TMAIL_DOMAIN_SUFFIX) else { return false }

        let base = String(trimmed.dropLast(TMAIL_DOMAIN_SUFFIX.count))
        guard !base.isEmpty else { return false }

        let range = NSRange(location: 0, length: base.utf16.count)
        return TMAIL_ALIAS_REGEX.firstMatch(in: base, options: [], range: range) != nil
    }

    public static func tmailAliasBase(_ value: String) -> String? {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard trimmed.hasSuffix(TMAIL_DOMAIN_SUFFIX) else { return nil }

        let base = String(trimmed.dropLast(TMAIL_DOMAIN_SUFFIX.count))
        guard !base.isEmpty else { return nil }

        let range = NSRange(location: 0, length: base.utf16.count)
        return TMAIL_ALIAS_REGEX.firstMatch(in: base, options: [], range: range) != nil ? base : nil
    }

    /// Mailbox carried by a TMail share QR: `https://<any-host>/share/<url-encoded mailbox>`.
    /// The host is not checked — dev/staging/prod hosts all differ.
    public static func parseShareQr(_ raw: String) -> String? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let markerRange = trimmed.range(of: TMAIL_SHARE_PATH) else { return nil }

        // `#` and `@` of a web3 mailbox are percent-encoded in the QR, so cut the tail off before decoding.
        let encoded = trimmed[markerRange.upperBound...]
            .prefix { $0 != "?" && $0 != "#" && $0 != "/" }
        guard let mailbox = String(encoded).removingPercentEncoding?
            .trimmingCharacters(in: .whitespacesAndNewlines),
              mailbox.contains("@") else { return nil }

        return mailbox
    }

    /// Bare local-part alias (no `.` / `@`) that can be resolved as `@tmail.ton` then `.ton` DNS.
    public static func isBareTonAlias(_ value: String) -> Bool {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard !trimmed.isEmpty, !trimmed.contains("."), !trimmed.contains("@") else { return false }

        let range = NSRange(location: 0, length: trimmed.utf16.count)
        return TMAIL_ALIAS_REGEX.firstMatch(in: trimmed, options: [], range: range) != nil
    }
}

File: mobile/android/air/SubModules/WalletContext/src/main/java/app/twallet/air/walletcontext/helpers/helpers/TmailHelpers.kt
package app.twallet.air.walletcontext.helpers

import android.net.Uri

// Tmail alias detection helpers. Kept separate from DNSHelpers so the two detectors don't mix.
object TmailHelpers {

    private const val TMAIL_DOMAIN_SUFFIX = "@tmail.ton"
    private const val TMAIL_SHARE_PATH = "/share/"
    private val TMAIL_ALIAS_REGEX = Regex("^[a-z0-9]([-_+a-z0-9]{0,62}[a-z0-9])?$", RegexOption.IGNORE_CASE)

    fun isTmailAlias(value: String): Boolean {
        val trimmed = value.trim().lowercase()
        if (!trimmed.endsWith(TMAIL_DOMAIN_SUFFIX)) return false

        val base = trimmed.dropLast(TMAIL_DOMAIN_SUFFIX.length)
        if (base.isEmpty()) return false

        return TMAIL_ALIAS_REGEX.matches(base)
    }

    fun tmailAliasBase(value: String): String? {
        val trimmed = value.trim().lowercase()
        if (!trimmed.endsWith(TMAIL_DOMAIN_SUFFIX)) return null

        val base = trimmed.dropLast(TMAIL_DOMAIN_SUFFIX.length)
        if (base.isEmpty()) return null

        return if (TMAIL_ALIAS_REGEX.matches(base)) base else null
    }

    /**
     * Mailbox carried by a TMail share QR: `https://<any-host>/share/<url-encoded mailbox>`.
     * The host is not checked — dev/staging/prod hosts all differ.
     */
    fun parseShareQr(raw: String): String? {
        val trimmed = raw.trim()
        val markerIndex = trimmed.indexOf(TMAIL_SHARE_PATH)
        if (markerIndex < 0) return null

        // `#` and `@` of a web3 mailbox are percent-encoded in the QR, so cut the tail off before decoding.
        val encoded = trimmed.substring(markerIndex + TMAIL_SHARE_PATH.length)
            .substringBefore('?').substringBefore('#').substringBefore('/')
        // Uri.decode, not URLDecoder — the latter would turn a `+` in the alias into a space.
        val mailbox = Uri.decode(encoded)?.trim() ?: return null

        return if (mailbox.contains('@')) mailbox else null
    }

    /** Bare local-part alias (no `.` / `@`) that can be resolved as `@tmail.ton` then `.ton` DNS. */
    fun isBareTonAlias(value: String): Boolean {
        val trimmed = value.trim().lowercase()
        if (trimmed.isEmpty() || trimmed.contains('.') || trimmed.contains('@')) return false

        return TMAIL_ALIAS_REGEX.matches(trimmed)
    }
}
